File: app/server/internal/model/guild.go
package model

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// Guild 代表一个语音服务器（类 Discord Server/Guild）。
type Guild struct {
	ID          uint      `gorm:"primaryKey" json:"id"`
	UUID        string    `gorm:"type:uuid;uniqueIndex" json:"uuid"`
	Name        string    `gorm:"size:100;not null" json:"name"`
	IconURL     string    `gorm:"size:512" json:"icon_url"`
	Description string    `gorm:"size:500" json:"description"`
	OwnerUUID   string    `gorm:"type:uuid;index;not null" json:"owner_uuid"`
	InviteCode  string    `gorm:"size:32;uniqueIndex" json:"invite_code"`
	MaxRooms    uint      `gorm:"default:0" json:"max_rooms"`
	IsPublic    bool      `gorm:"default:false" json:"is_public"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

func (g *Guild) TableName() string {
	return "guilds"
}

func (g *Guild) BeforeCreate(_ *gorm.DB) error {
	if g.UUID == "" {
		g.UUID = uuid.New().String()
	}
	if g.InviteCode == "" {
		g.InviteCode = generateInviteCode()
	}
	return nil
}

// GuildMember 用户-Guild 多对多关系。
type GuildMember struct {
	ID        uint      `gorm:"primaryKey" json:"id"`
	GuildUUID string    `gorm:"type:uuid;index:idx_guild_member,priority:1;not null" json:"guild_uuid"`
	UserUUID  string    `gorm:"type:uuid;index:idx_guild_member,priority:2;not null" json:"user_uuid"`
	Nickname  string    `gorm:"size:64" json:"nickname"`
	RoleName  string    `gorm:"size:32;default:member" json:"role_name"`
	JoinedAt  time.Time `json:"joined_at"`
}

func (GuildMember) TableName() string {
	return "guild_members"
}

// generateInviteCode 生成 8 字符随机邀请码。
func generateInviteCode() string {
	const charset = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"
	u := uuid.New()
	b := make([]byte, 8)
	for i := range b {
		b[i] = charset[u.ID()%uint32(len(charset))]
	}
	return string(b)
}

File: app/web/src/api/guild.ts
import type { AxiosResponse } from "axios";
import type { Result } from "./apiClient";
import apiClient from "./apiClient";

export interface Guild {
	id: number;
	uuid: string;
	name: string;
	icon_url: string;
	description: string;
	owner_uuid: string;
	invite_code: string;
	max_rooms: number;
	is_public: boolean;
	created_at: string;
}

export interface GuildMember {
	id: number;
	guild_uuid: string;
	user_uuid: string;
	nickname: string;
	role_name: string;
	joined_at: string;
}

function unwrap<T>(res: AxiosResponse<Result<T>>): T {
	const data = res.data.data;
	if (data == null) throw new Error("guild data is missing");
	return data;
}

export async function createGuild(data: {
	name: string;
	description?: string;
	is_public?: boolean;
}): Promise<Guild> {
	const res = await apiClient.post({
		url: "/api/v1/guild/create",
		data,
	});
	return unwrap<Guild>(res as AxiosResponse<Result<Guild>>);
}

export async function getGuild(uuid: string): Promise<Guild> {
	const res = await apiClient.post({
		url: "/api/v1/guild/get",
		data: { uuid },
	});
	return unwrap<Guild>(res as AxiosResponse<Result<Guild>>);
}

export async function listGuilds(
	page = 1,
	pageSize = 20,
): Promise<{ guilds: Guild[]; total: number }> {
	const res = await apiClient.post({
		url: "/api/v1/guild/list",
		data: { page, page_size: pageSize },
	});
	return unwrap<{ guilds: Guild[]; total: number }>(
		res as AxiosResponse<Result<{ guilds: Guild[]; total: number }>>,
	);
}

export async function listPublicGuilds(
	page = 1,
	pageSize = 20,
	keyword?: string,
): Promise<{ guilds: Guild[]; total: number }> {
	const data: { page: number; page_size: number; keyword?: string } = {
		page,
		page_size: pageSize,
	};
	if (keyword?.trim()) {
		data.keyword = keyword.trim();
	}
	const res = await apiClient.post({
		url: "/api/v1/guild/list-public",
		data,
	});
	return unwrap<{ guilds: Guild[]; total: number }>(
		res as AxiosResponse<Result<{ guilds: Guild[]; total: number }>>,
	);
}

export async function myGuilds(): Promise<string[]> {
	const res = await apiClient.post({
		url: "/api/v1/guild/my-guilds",
		data: {},
	});
	const data = (res as AxiosResponse<Result<{ guild_uuids: string[] }>>).data
		.data;
	return data?.guild_uuids ?? [];
}

export async function previewGuildInvite(inviteCode: string): Promise<Guild> {
	const res = await apiClient.post({
		url: "/api/v1/guild/preview",
		data: { invite_code: inviteCode },
	});
	return unwrap<Guild>(res as AxiosResponse<Result<Guild>>);
}

export async function joinGuild(inviteCode: string): Promise<Guild> {
	const res = await apiClient.post({
		url: "/api/v1/guild/join",
		data: { invite_code: inviteCode },
	});
	return unwrap<Guild>(res as AxiosResponse<Result<Guild>>);
}

export async function leaveGuild(uuid: string): Promise<void> {
	await apiClient.post({
		url: "/api/v1/guild/leave",
		data: { uuid },
	});
}

export async function updateGuild(data: {
	uuid: string;
	name?: string;
	description?: string;
	icon_url?: string;
	is_public?: boolean;
	max_rooms?: number;
}): Promise<Guild> {
	const res = await apiClient.post({
		url: "/api/v1/guild/update",
		data,
	});
	return unwrap<Guild>(res as AxiosResponse<Result<Guild>>);
}

export async function deleteGuild(uuid: string): Promise<void> {
	await apiClient.post({
		url: "/api/v1/guild/delete",
		data: { uuid },
	});
}

export async function kickGuildMember(
	guildUUID: string,
	userUUID: string,
): Promise<void> {
	await apiClient.post({
		url: "/api/v1/guild/kick",
		data: { guild_uuid: guildUUID, user_uuid: userUUID },
	});
}

export async function guildMembers(guildUUID: string): Promise<GuildMember[]> {
	const res = await apiClient.post({
		url: "/api/v1/guild/members",
		data: { guild_uuid: guildUUID },
	});
	const data = (res as AxiosResponse<Result<{ members: GuildMember[] }>>).data
		.data;
	return data?.members ?? [];
}
